FormalisingMathematics2025/Section03functions/Sheet1.lean

```diff
@@ -48,44 +43,34 @@ example : ℕ := 2
 -- # But this isn't maths!
 
 -- The type `Prop` contains `Prop`ositions...
-
 #check 2 + 2 = 4
 #check rexp 1 < π
-
 #check 2 + 2 = 5
 #check Irrational (rexp 1 + π)
 #check myFavouriteNumber = yourFavouriteNumber
-
 def MyDifficultProposition : Prop := ∀ n : ℕ, ∃ p, n ≤ p ∧ Prime p ∧ Prime (p + 2)
 def MyEasyProposition : Prop := ∀ n : ℕ, ∃ p, n ≤ p ∧ Prime p ∧ Prime (p + 2) ∧ Prime (p + 4)
-def MyVeryEasyProposition : Prop := ∀ n : ℕ, ∃ p, n ≤ p
 
 -- Key! If `p : Prop`, an expression of type `p` is a proof of `p`.
-
 example : 2 + 2 = 4 := rfl
 example : 2 + 2 ≠ 5 := by simp
 example : ∀ n : ℕ, 2 ≤ n → ∃ x y z : ℕ, 4 * x * y * z = n * (x * y + x * z + y * z) := sorry
--- Erdős-Strauss conjecture
-
-example (n : ℕ) (hn : 2 ≤ n) :
-  ∃ x y z : ℕ, 4 * x * y * z = n * (x * y + x * z + y * z) := sorry
+-- Erdős-Strauss conjecture. We're not going to prove it now (unless you know how?)
 
 -- # How can we make these expressions?
 
 -- Simple proof terms
 example : True := trivial
 example : 2 = 2 := rfl
 example (a b : ℕ) : a + b = b + a := Nat.add_comm a b
-
 example (a b : ℕ) : a * b = b * a := Nat.mul_comm a b
 
-theorem my_proof : MyVeryEasyProposition := fun n => ⟨n, le_rfl⟩
-
+def MyVeryEasyProposition : Prop := 2 = 2
+def my_proof : MyVeryEasyProposition := rfl
 #check MyVeryEasyProposition
 #check my_proof
 -- my proposition "has type Proposition", or "is a proposition"
--- my proof "has type my proposition", or "has type ∀ n : ℕ, ∃ p, n ≤ p",
---    or "is a proof of ∀ n : ℕ, ∃ p, n ≤ p"
+-- my proof "has type my proposition", or "has type 2 = 2", or "is a proof of 2 = 2"
 
 -- But just proof terms get ugly...
 example (a b : ℕ) : a + a * b = (b + 1) * a :=
```

Comment:
Now by this point I think we have the story flowing nicely:

- First, we've shown expressions have types
- Then we taught to give them names via `def`
- Then we taught propositions are of type `Prop`
- Then we taught that proofs are of type of their propositions
- Then we taught that it's easier to write proof via `by` and tactics

Each concept builds on the other, each step is introduced on its own with no sudden jumps introducing two things at once.